DebuggingFramework.md

```diff
@@ -0,0 +1,139 @@
+modules.
+
+## Introduction
+
+There are several entities involved in debugging a source program compiled into
+WebAssembly:
+
+- the compiled code being executed, called the _debuggee_
+- the browser engine executing the debuggee
+- the browser devtools
+- the source-program listing, usually split into multiple source files
+- the description of how the debuggee code maps to the source program and vice
+  versa (e.g., WebAssembly code locations to source code locations, source
+  variables to WebAssembly memory, WebAssembly call stack to source functions,
+  etc.), called the _debug info_
+
+These entities carry information and interact in particular ways to enable a
+user to perform debugging actions at the source-code level.  For example, the
+browser devtools offer a UI allowing the user to step through the source code,
+requiring partial execution of the debuggee chunk corresponding to each step,
+and displaying the source file containing the stepped-through program.
+
+Specifying the debugging framework means describing how these entities interact,
+what they need from each other, and how they access it; all in order to execute
+the user's commands.
+
+## Debugger as a Procedural Interface
+
+A crucial element of this design proposal is to omit specifying a particular
+format for the debug info.  Instead, we introduce a separate entity we call the
+debugger; its job is to act as a procedural intermediary between the devtools
+and the debug info.  Devtools execute user actions by leveraging debugger
+methods we will specify here.
+
+Thus any debug-info format is acceptable, as long as the debugger can correctly
+interpret it.  The format is, in effect, a convention between the front-end
+parsing the source and the debugger.
+
+What the design _does_ specify is two interaction protocols:
+
+1. between the devtools and the debugger, and
+2. between the debugger and the wasm execution engine
+
+## Action Flow
+
+For reference, the entities are charted in the figure below:
+
+![Entities making up the debug framework](wasm-debug-chart.png)
+
+(The figure uses the conventional phrase "wasm module" to refer to a debuggee.
+For now, we overlook the complexities of multiple wasm modules or mixing JS and
+wasm on the same call stack.)
+
+Black arrows in the figure indicate the action flow and timeline:
+
+1. When the page is first rendered, the debuggee is loaded into the browser and
+   sent to the wasm execution engine.
+2. The user interacts with the browser's devtools UI (eg, requesting a source
+   listing for a function to set a breakpoint).
+3. Devtools call on the debugger to execute operation(s) necessary to fulfilling
+   the user's request.  After going through steps 4 and/or 5, the debugger
+   returns a result for the devtools to display.
+4. The debugger consults the debug info and/or source files to obtain the proper
+   mapping between the source code and the wasm.
```

Comment:
Yeah, I felt the extra arrow would clutter the picture without adding much information value.  I glommed the debug info and source files rectangles together to suggest that they act in unison.